File: methlib/container/src/main/java/com/ogms/dge/container/modules/method/processor/CommandProcessor.java
package com.ogms.dge.container.modules.method.processor;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.ogms.dge.container.modules.method.dto.CmdContextDto;
import com.ogms.dge.container.modules.method.dto.CmdDto;
import com.ogms.dge.container.modules.method.entity.MethodEntity;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.io.File;
import java.io.IOException;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;

/**
 * @name: CommandProcessor
 * @description: TODO
 * @author: Lingkai Shi
 * @date: 12/19/2024 3:50 PM
 * @version: 1.0
 */
@Component
public class CommandProcessor {
    @Value("${container.method.wd}")
    private String method_wd;

    @Value("${container.data.fd}")
    private String data_fd;

    @Value("${container.data.td}")
    private String data_td;

    @Value("${container.data.server}")
    private String data_server;

    @Value("${container.method.pd}")
    private String method_pd;

    @Value("${container.env.py}")
    private String env_py;

    @Autowired
    private ParameterHandlerFactory parameterHandlerFactory;

    private ObjectMapper objectMapper = new ObjectMapper();

    public CmdDto getCmd(Long userId, MethodEntity method, String serviceBasePath, String serviceUuid, Map<String, Object> reqParams, Date startTime, Boolean isExternalCall) throws JsonProcessingException {
        List<Map<String, Object>> paramSpecsList = objectMapper.readValue(method.getParams(), new TypeReference<List<Map<String, Object>>>() {
        });
        StringBuilder cmdBuilder = new StringBuilder();

        File wdFolder = new File(method_wd + UUID.randomUUID().toString());
        wdFolder.mkdirs(); // 如果目录已存在，不会重复创建

        if (method.getId() > 478L) {
            // 非Whitebox方法
            cmdBuilder.append("cd /d \"").append(wdFolder.getAbsolutePath()).append("\" && ");
        }
        cmdBuilder.append(getToolCmd(method, method.getUuid())).append(" ");
        // 针对 Whitebox，添加方法名和临时工作目录
        if (method.getId() < 479L) {
            cmdBuilder.append("-r=").append(method.getName()).append(" ");
            cmdBuilder.append("--wd=").append(wdFolder.getAbsolutePath()).append(" ");
        }
        // 针对 Saga 方法，添加 toolIndex
        if (method.getId() > 581L && method.getId() < 1342L) {
            cmdBuilder.append(method.getCategory()).append(" ");
        }

        // 初始化多个变量，方便后期记录，命令执行后处理
        CmdDto cmdDto = new CmdDto();
        CmdContextDto cmdContextDto = new CmdContextDto();
        boolean isSuccess = true;
        try {
            // 以下开始参数处理
            for (int i = 0; i < paramSpecsList.size(); i++) {
                if (!reqParams.containsKey("val" + i)) {
                    // 用户没有操作此项参数
                    continue;
                }
                Map<String, Object> paramSpecs = paramSpecsList.get(i);
                // 先获取参数输入标识flag,获取 Flags 数组的第一个元素
                processFlags(cmdBuilder, paramSpecs);

                cmdContextDto = new CmdContextDto(cmdBuilder, paramSpecs, wdFolder, cmdDto, serviceBasePath, serviceUuid, userId);

                // 获取 parameter_type 属性并分类处理
                Object parameterTypeObj = paramSpecs.get("parameter_type");
                ParameterHandler handler = parameterHandlerFactory.getHandler(parameterTypeObj);
                handler.parse(parameterTypeObj, reqParams.get("val" + i), i, cmdContextDto, isExternalCall);
            }
        } catch (Exception e) {
            isSuccess = false;
            // addMethodLog(method.getId(), userId, wdFolder.getName(), null,serviceUuid==null?0:1, null, 1, "Parameters Error", startTime);
        }
        if (!isSuccess) {
            return null;
        }
        cmdDto = cmdContextDto.getCmdDto();
        cmdDto.cmd = cmdContextDto.getCmdBuilder().toString();
        // cmdDto.fileFrontNameList = cmdDto.getOutputFileRealNameList();
        if (!isExternalCall)
            cmdDto.outputFilePath = Paths.get(data_fd, Objects.toString(userId)).toString(); // TODO userId
        cmdDto.tmpFilePath = wdFolder.getAbsolutePath();
        return cmdDto;
    }

    private void processFlags(StringBuilder cmdBuilder, Map<String, Object> paramSpecs) {
        Object flagsObj = paramSpecs.get("Flags");
        if (flagsObj instanceof List<?>) {
            List<?> flagsList = (List<?>) flagsObj;
            if (!flagsList.isEmpty()) {
                cmdBuilder.append(flagsList.get(0)).append(" ");
            } else {
                cmdBuilder.append(" ");
            }
        }
    }

    private String getToolCmd(MethodEntity method, String uuid) {
        // 获取exe地址，TODO 需要确保环境路径无空格
        String path = "";
        if (method.getId() < 479L) {
            Path directoryPath = Paths.get(method_pd + uuid);
            try (DirectoryStream<Path> stream = Files.newDirectoryStream(directoryPath, "*.exe")) {
                for (Path entry : stream) {
                    path = entry.toAbsolutePath().toString();
                    break;
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else if (method.getExecution().equals("py")) {
            Path directoryPath = Paths.get(method_pd + uuid);
            try (DirectoryStream<Path> stream = Files.newDirectoryStream(directoryPath, "*.py")) {
                for (Path entry : stream) {
                    path = Paths.get(env_py + "python.exe").toAbsolutePath() + " " + entry.toAbsolutePath().toString() + " "; // python test.py
                    break;
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else if (method.getExecution().equals("exe")) {
            Path directoryPath = Paths.get(method_pd + uuid);
            try (DirectoryStream<Path> stream = Files.newDirectoryStream(directoryPath, "*.exe")) {
                for (Path entry : stream) {
                    path = entry.toAbsolutePath().toString() + " ";
                    break;
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else if (method.getExecution().equals("jar")) {
            Path directoryPath = Paths.get(method_pd + uuid);
            try (DirectoryStream<Path> stream = Files.newDirectoryStream(directoryPath, "*.jar")) {
                for (Path entry : stream) {
                    path = "java -jar " + entry.toAbsolutePath().toString() + " ";
                    break;
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return path;
    }
}

File: modelServer/dataProcessing/model/methlib/parsers/command_processor.py
# 全局工厂初始化 (模拟 dataProcessing.model.methlib.__init__.py)
from datetime import datetime
import json
from pathlib import Path
from typing import Any, Dict, List
import uuid
from dataProcessing.model.methlib.parsers.factory import ParameterHandlerFactory
from dataProcessing.model.methlib.schemas.command import CmdContext, CmdDto
from dataProcessing.model.methlib.schemas.method import MethodEntity
from dataProcessing.model.methlib import init_handler_factory
from dataProcessing.config import current_config as CONFIG


class CommandProcessor:
    """核心命令处理器：负责根据方法和请求参数构建最终的命令行字符串"""
    def __init__(self):
        # 模拟 @Autowired 注入
        self.parameter_handler_factory = init_handler_factory()

        # 模拟 @Value 配置
        self.method_wd = CONFIG.METHOD_WD
        self.data_fd = CONFIG.DATA_FD
        self.method_pd = CONFIG.METHOD_PD

    def get_cmd(self, user_id: str, method: MethodEntity, service_base_path: str, service_uuid: str, req_params: Dict[str, Any], start_time: datetime, is_external_call: bool) -> CmdDto | None:
        
        # 1. 解析参数规格
        param_specs_list = method.params

        cmd_builder: List[str] = []
        is_success = True
        
        # 2. 创建工作目录 (wdFolder)
        wd_folder_path = Path(self.method_wd) / str(uuid.uuid4())
        try:
            wd_folder_path.mkdir(parents=True, exist_ok=True)
        except OSError as e:
            print(f"Error creating working directory: {e}")
            return None 

        # 3. 构建基础命令前缀
        if method.id > 478:
            # 非Whitebox方法: 添加 cd 命令
            cmd_builder.append(f"cd /d \"{wd_folder_path.absolute().as_posix()}\" && ")
        
        cmd_builder.append(self._get_tool_cmd(method, method.uuid))
        
        # 针对 Whitebox, 添加方法名和临时工作目录
        if method.id < 479:
            cmd_builder.append(f"-r={method.name} ")
            cmd_builder.append(f"--wd=\"{wd_folder_path.absolute().as_posix()}\" ")
            
        # 针对 Saga 方法，添加 toolIndex
        if 581 < method.id < 1342:
            cmd_builder.append(f"{method.category} ")

        # 4. 初始化上下文
        cmd_dto = CmdDto()
        cmd_context = CmdContext(
            cmd_builder=cmd_builder, 
            param_specs={},
            wd_folder=wd_folder_path, 
            cmd_dto=cmd_dto, 
            service_base_path=service_base_path, 
            service_uuid=service_uuid, 
            user_id=user_id
        )

        # 5. 参数处理核心循环
        try:
            for i, param_specs in enumerate(param_specs_list):
                val_key = f"val{i}"
                if val_key not in req_params:
                    # 用户没有操作此项参数
                    continue
                
                # 5.1 处理 Flags (添加参数标识符)
                self._process_flags(cmd_builder, param_specs)

                # 5.2 更新上下文的 paramSpecs
                cmd_context.param_specs = param_specs

                # 5.3 获取处理器并解析
                parameter_type_obj = param_specs.get("parameter_type")
                
                handler = self.parameter_handler_factory.get_handler(parameter_type_obj)
                
                handler.parse(
                    parameter_type=parameter_type_obj,
                    raw_value=req_params.get(val_key),
                    val_index=i,
                    context=cmd_context,
                    is_external_call=is_external_call
                )

        except Exception as e:
            is_success = False
            print(f"Parameters Error occurred: {e}")
            
        if not is_success:
            return None

        # 6. 整理 CmdDto 输出
        cmd_dto = cmd_context.cmd_dto
        cmd_dto.cmd = "".join(cmd_context.cmd_builder).strip()
        
        # 路径设置
        if not is_external_call:
            pass
            # cmd_dto.output_file_path = str(Path(self.data_fd) / str(user_id))
        
        cmd_dto.tmp_file_path = str(wd_folder_path.absolute().as_posix())
        
        return cmd_dto

    def _process_flags(self, cmd_builder: List[str], param_specs: Dict[str, Any]):
        """
        处理参数输入标识 flags。
        严格复刻 Java processFlags 逻辑：只处理 List 类型，空 List 也要加一个空格。
        """
        flags_obj = param_specs.get("Flags")
        if isinstance(flags_obj, list):
            if flags_obj:
                cmd_builder.append(f"{flags_obj[0]} ")
            else:
                # 严格匹配 Java 行为：空列表也加一个空格
                cmd_builder.append(" ")
        # 如果不是 list (包括 None)，则不操作，匹配 Java 行为。

    def _get_tool_cmd(self, method: MethodEntity, uuid: str) -> str:
        """获取工具的可执行地址和前缀命令，匹配 Java getToolCmd 逻辑"""
        
        path_prefix = Path(self.method_pd) / uuid

        # 特殊处理 Whitebox 方法 (id < 479L)，直接写死 whitebox_tools 路径
        if method.id < 479:
            # 构造完整可执行路径
            tool_path = (path_prefix / "whitebox_tools").absolute().as_posix()
            # 注意这里不加引号，因为原逻辑 EXE 模式不加引号
            return f"{tool_path} "

        # === 下面保持原逻辑 ===
        path = ""
        extension_map = {
            "exe": [("*.exe", "", " ")],
            "jar": [("*.jar", "java -jar ", " ")],
        }

        execution_type = method.execution

        if execution_type == "py":
            tool_info = extension_map.get("py")[0]
        elif execution_type == "exe":
            tool_info = extension_map.get("exe")[0]
        elif execution_type == "jar":
            tool_info = extension_map.get("jar")[0]
        else:
            return ""

        glob_pattern, prefix, suffix = tool_info
        tool_files = list(path_prefix.glob(glob_pattern))
        
        if tool_files:
            tool_path = tool_files[0].absolute().as_posix()
            if execution_type == "py" or execution_type == "jar":
                 path = f"{prefix}\"{tool_path}\"{suffix}"
            else:
                 path = f"{prefix}{tool_path}{suffix}"
        
        return path

